File: example/cpp/tests/result.cpp
#include <iostream>
#include "../ResultOpaque.hpp"
#include "assert.hpp"

int main(int argc, char *argv[]) {
    ResultOpaque r;
    ResultOpaque r2 = ResultOpaque::new_(5).ok().value();
    r2.assert_integer(5);
    auto foo_err = ResultOpaque::new_failing_foo();
    auto foo = foo_err.err().value();
    simple_assert_eq("foo error", (int)foo, (int)ErrorEnum::Foo);

    auto bar_err = ResultOpaque::new_failing_bar();
    auto bar = bar_err.err().value();
    simple_assert_eq("bar error", (int)bar, (int)ErrorEnum::Bar);

    auto unit_err = ResultOpaque::new_failing_unit();
    simple_assert("unit error", unit_err.is_err())

    auto struc_err = ResultOpaque::new_failing_struct(109);
    auto struc = struc_err.err().value();
    simple_assert_eq("struct error", struc.i, 109);

    auto in_err = ResultOpaque::new_in_err(198);
    auto in_err_val = in_err.err().value();
    in_err_val.assert_integer(198);

    auto in_enum_err = ResultOpaque::new_in_enum_err(989);
    auto in_enum_err_val = in_enum_err.err().value();
    in_enum_err_val.assert_integer(989);
}

File: example/src/feature_tests/result.rs
#[diplomat::bridge]
pub mod ffi {

    use diplomat_runtime::DiplomatResult;

    #[diplomat::opaque]
    pub struct ResultOpaque(i32);

    #[derive(PartialEq, Eq, Debug)]
    pub enum ErrorEnum {
        Foo,
        Bar,
    }

    #[derive(Debug)]
    pub struct ErrorStruct {
        i: i32,
    }
    impl ResultOpaque {
        pub fn new(i: i32) -> DiplomatResult<Box<ResultOpaque>, ErrorEnum> {
            Ok(Box::new(ResultOpaque(i))).into()
        }

        pub fn new_failing_foo() -> DiplomatResult<Box<ResultOpaque>, ErrorEnum> {
            Err(ErrorEnum::Foo).into()
        }

        pub fn new_failing_bar() -> DiplomatResult<Box<ResultOpaque>, ErrorEnum> {
            Err(ErrorEnum::Bar).into()
        }

        pub fn new_failing_unit() -> DiplomatResult<Box<ResultOpaque>, ()> {
            Err(()).into()
        }

        pub fn new_failing_struct(i: i32) -> DiplomatResult<Box<ResultOpaque>, ErrorStruct> {
            Err(ErrorStruct { i }).into()
        }

        pub fn new_in_err(i: i32) -> DiplomatResult<(), Box<ResultOpaque>> {
            Err(Box::new(ResultOpaque(i))).into()
        }

        pub fn new_in_enum_err(i: i32) -> DiplomatResult<ErrorEnum, Box<ResultOpaque>> {
            Err(Box::new(ResultOpaque(i))).into()
        }

        pub fn assert_integer(&self, i: i32) {
            assert_eq!(i, self.0);
        }
    }
}
